prov/sm2/src/sm2_coordination.c

```diff
@@ -0,0 +1,675 @@
+				// use this region for as long as the file exists
+				FI_WARN(&sm2_prov, FI_LOG_AV,
+					"Found region at AV[%d] that did not shut down "
+					"correctly, marking it as a zombie never to be "
+					"used again!\n",
+					jentry);
+				strncpy(entries[jentry].ep_name, ZOMBIE_ALLOCATION_NAME,
+					SM2_NAME_MAX);
+				goto retry_lookup;
+			}
+		}
+
+		if (!self) {
+			if (!pid_lives(abs(entries[jentry].pid))) {
+				entries[jentry].pid = 0;
+			}
+			// Someone else allocated the entry for us
+			goto found;
+		}
+
+		if (entries[jentry].pid <= 0) {
+			if (!pid_lives(abs(entries[jentry].pid))) {
+				FI_WARN(
+				    &sm2_prov, FI_LOG_AV,
+				    "during sm2 allocation of space for endpoint named %s"
+				    " pid %d pre-allocated space at AV[%d] and then "
+				    "died!\n",
+				    name, -entries[jentry].pid, jentry);
+			}
+			goto found;
+		}
+
+		FI_WARN(&sm2_prov, FI_LOG_AV,
+			"During sm2 allocation of space for endpoint named %s "
+			"an existing conflicting address was found at AV[%d]\n",
+			name, jentry);
+
+		if (!pid_lives(entries[jentry].pid)) {
+			FI_WARN(&sm2_prov, FI_LOG_AV,
+				"The pid which allocated the conflicting AV is dead.  "
+				"Reclaiming as our own.\n");
+			// it is possible that EP's referencing this region are
+			// still alive... don't know how to check (they likely
+			// died if PID died)
+			goto found;
+		}
+		else {
+			FI_WARN(&sm2_prov, FI_LOG_AV,
+				"ERROR: The endpoint (pid: %d) with conflicting address "
+				"%s is still alive.\n",
+				entries[jentry].pid, name);
+			return -FI_EADDRINUSE;
+		}
+	}
+
+	/* fine, we could not find the entry, so now look for an empty slot */
+	for (jentry = 0; jentry < header->ep_enumerations_max; jentry++) {
+		peer_pid = entries[jentry].pid;
+		if (peer_pid == 0)
+			goto found;
+		else if (peer_pid < 0) {
+			// A third peer might have entered this address into their AV,
+			// and there is no current way to check this... need to keep this
+			// entry in the file until we clean up
+			continue;
+		}
+		else {
+			if (!pid_lives(peer_pid)) {
+				sm2_coordinator_extend_for_entry(map, jentry);
+				entries = sm2_mmap_entries(map);
+				header = (void *)map->base;
+				struct sm2_region *peer_region =
+				    sm2_mmap_ep_region(map, jentry);
+
+				if (entries[jentry].startup_ready &&
+				    smr_freestack_isfull(sm2_free_stack(peer_region))) {
+					/* we found a slot with a dead PID and the
+					 * freestack is full */
+					entries[jentry].pid = 0;
+					goto found;
+				};
+			}
+		}
+	}
+
+	fprintf(stderr,
+		"No available entries were found in the coordination file, all %d were "
+		"used\n",
+		header->ep_enumerations_max);
+	return -FI_EAVAIL;
+
+found:
+
+	if (self) {
+		entries[jentry].startup_ready = 0;
+		atomic_wmb();
+		entries[jentry].pid = pid;
+	}
+
+	if (!self && entries[jentry].pid == 0) {
+		entries[jentry].startup_ready = 0;
+		atomic_wmb();
+		entries[jentry].pid = -pid;
```

Comment:
The negative PID means that a processes reserved a spot in `struct sm2_ep_allocation_entry`  for a process with an ADDR that isn't up yet.  This happens because on fi_av_insert(), we need to immediately return the SHM_ID (the requirement may lossen if we use util_av).  

Example:
Luke comes alive (takes spot 1)
Luke fi_av_inserts() Alexia's address, pre-allocates a spot for her, and returns spot 1
seth comes allve (takes spot 2)
Seth fi_av_inserts() Alexia's address, finds Alexia's spot and returns spot 1
Alexia comes alive and realizes that spot 1 has been, pre-allocated for her, fills it out, and marks her region initialized, and ready to receive messages